### PR TITLE
ci: lua language server and Makefile

### PR DESCRIPTION
**Pull request by alex-courtis**
Adds lua-language-server check

Add Makefile

Document contributing.

Rework CI.

### CODE REVIEW

**Comment by alex-courtis** [comment]
I reckon we should wait until #2555 as it may make @Akmadan23 's life difficult.

File: scripts/lls-check [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+DIR_LLS="/tmp/lls"
+
+# clear output
+rm -rf "${DIR_OUT}"
+mkdir "${DIR_OUT}"
+
+if [ ! -d "${DIR_LLS}" ]; then
+	# download lua-language-server binaries
+	mkdir -p "${DIR_LLS}"
+	curl -L "https://github.com/LuaLS/lua-language-server/releases/download/${VER_LLS}/lua-language-server-${VER_LLS}-linux-x64.tar.gz" | tar zx --directory "${DIR_LLS}"
```
Comment (alex-courtis) [inline scripts/lls-check:20]: Ubuntu doesn't have a package and there don't seem to be any other means to source and run this.

File: .luarc.json [hunk 2]
```diff
@@ -1,6 +1,12 @@
...
   "runtime.version" : "Lua 5.1",
+  "workspace": {
+    "library": [
+      "$VIMRUNTIME",
+      "${3rd}/luv/library"
+    ]
+  },
   "diagnostics": {
     "globals": [
       "vim"
```
Comment (alex-courtis) [inline .luarc.json:12]: We can't remove this just yet as there are some warnings, although they may be spurious.

File: scripts/luals-check [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+DIR_LLS="/tmp/lls"
+
+# clear output
+rm -rf "${DIR_OUT}"
+mkdir "${DIR_OUT}"
+
+if [ ! -d "${DIR_LLS}" ]; then
+	# download lua-language-server binaries
+	mkdir -p "${DIR_LLS}"
+	curl -L "https://github.com/LuaLS/lua-language-server/releases/download/${VER_LLS}/lua-language-server-${VER_LLS}-linux-x64.tar.gz" | tar zx --directory "${DIR_LLS}"
```
Comment (alex-courtis) [inline scripts/luals-check:20]: That's a bad idea; that binary is only appropriate for CI.

Users will need to install the language server themselves to be able to run this.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -4,31 +4,51 @@ Thank you for contributing.
...
+## Check
+
+[luals](https://luals.github.io) check is run with:
+
+```sh
+scripts/luals-check.sh
+```
+
+Requires `lua-language-server` on your path.
+
```
Comment (alex-courtis) [inline CONTRIBUTING.md:38]: Added to wiki as well: https://github.com/nvim-tree/nvim-tree.lua/wiki/Development#lua-language-server

**Comment by alex-courtis** [comment]
Thanks to @Akmadan23 we now have a clean codebase and can always run this.

File: .luarc.json [hunk 5]
```diff
@@ -1,7 +1,14 @@
 {
   "$schema": "https://raw.githubusercontent.com/sumneko/vscode-lua/master/setting/schema.json",
   "runtime.version" : "Lua 5.1",
+  "workspace": {
+    "library": [
+      "$VIMRUNTIME/lua/vim/lsp",
+      "${3rd}/luv/library"
```
Comment (alex-courtis) [inline .luarc.json:7]: Baby steps: just lsp and uv for now.

cmd, api and fn will take some time.

**Comment by alex-courtis** [comment]
This one's ready @Akmadan23 :

* run luals check via a script:
  * developer
  * CI - mandatory
* add lsp and libuv libraries

vim remains a global until all libraries can be added.

File: .luarc.json [hunk 6]
```diff
@@ -1,7 +1,14 @@
...
+    ]
+  },
   "diagnostics": {
+    "libraryFiles": "Disable",
```
Comment (alex-courtis) [inline .luarc.json:11]: There are many warnings in the libraries themselves.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -11,6 +11,12 @@ on:
 permissions:
   contents: read
 
+env:
+  LUA_VERSION: "5.1"
```
Comment (gegoune) [inline .github/workflows/ci.yml:15]: Maybe we should use matrix here instead of vars, will allow to easily extend to other versions and test them all.

**Review by gegoune — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 7 above)
Comment (gegoune) [reply .github/workflows/ci.yml:15]: Thank you!

Comment (alex-courtis) [reply .github/workflows/ci.yml]: Done.

Thrashed around for a while trying to use environment variables for matrix values but failed.

The matricies do work, setting `nvim_version: [ v0.8.3, v0.9.4 ]` worked but failed as the annotations are not present in 0.8.3.

Comment (alex-courtis) [reply .github/workflows/ci.yml]: I'd be most grateful if you could test the makefile as per `CONTRIBUTING.md`

**Pull request merged**